### ISSUE TITLE
`yescrypt` user passwords do not function

### ISSUE THREAD

**Issue by lourkeur** [NixOS/nixpkgs#147312]
### Describe the bug
When setting roots `hashedPassword` to a `yescrypt` hash, login becomes seemingly impossible even with the correct password. All else being equal, sha512crypt does not exhibit this problem.

### Steps To Reproduce
1. build this virtual machine: https://gist.github.com/lourkeur/771d190aeb026dba2a60897eab90de57
2. log in as root with password test
3. Womp womp

### Expected behavior
Access should be granted iff the password is correct.

### Additional context
This is a bit concerning given that yescrypt is now the `mkpasswd` default since #143661.

Should `pam` be built with the `withLibxcrypt` flag? If so how?

### Notify maintainers

For `pam`+`libxcrypt` interaction @dottedmag 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by Artturin** [comment]
Related https://github.com/NixOS/nixpkgs/issues/112371

**Comment by Artturin** [comment]
If yescrypt is now the default then I consider this a release blocker as we recommend mkpasswd in the manual

@amarshall

**Comment by lourkeur** [comment]
> If yescrypt is now the default then I consider this a release blocker as we recommend mkpasswd in the manual
> 
> @amarshall

That might not be true with `passwd`, but for declarative passwords yes. I think for installation, the manual is still imperative, correct?

**Comment by amarshall** [comment]
I can at least confirm that enabling `withLibxcrypt` in linux-pam allows login with yescrypt-hashed passwords. I’m not certain the broader implications of that change, but it’s probably a long-term goal. FWIW the [users-groups module presently warns if a yescrypt hash is configured](https://github.com/NixOS/nixpkgs/blob/27abb57b7fff5c4164de6ec3dd6840177e83a11e/nixos/modules/config/users-groups.nix#L658-L683).

Where in the manual does it recommend bare `mkpasswd`? In [NixOS installation](https://nixos.org/manual/nixos/stable/#sec-installation) it has `mkpasswd -m sha-512`.

**Comment by dottedmag** [comment]
Enabling libxcrypt in `linux-pam` makes at least `login` and SSH accept yescrypt-hashed passwods.

I'm not sure about the rest of the system: there is too much use of `libcrypt.so` (see the file attached to #112371) to be sure that some other packages don't try to access `/etc/shadow` directly.

**Comment by Artturin** [comment]
> I can at least confirm that enabling `withLibxcrypt` in linux-pam allows login with yescrypt-hashed passwords. I’m not certain the broader implications of that change, but it’s probably a long-term goal. FWIW the [users-groups module presently warns if a yescrypt hash is configured](https://github.com/NixOS/nixpkgs/blob/27abb57b7fff5c4164de6ec3dd6840177e83a11e/nixos/modules/config/users-groups.nix#L658-L683).
> 
> Where in the manual does it recommend bare `mkpasswd`? In [NixOS installation](https://nixos.org/manual/nixos/stable/#sec-installation) it has `mkpasswd -m sha-512`.

Ah nevermind then it's not as urgent now

**Comment by mweinelt** [comment]
Fixed in #195271.
